Product Add/Edit: Update price row and screen to support subscription

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #11135 
<!-- Id number of the GitHub issue this PR addresses. -->

## What
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the support for updating the subscription period for simple subscription products. This setting is accessible through the price setting screen as discussed in p1699462947949809-slack-C03L1NF1EA3.

Due to the size of the changes, the subscription sign-up fee and more unit tests will be added in a separate PR.

## How
- Added a new row for the subscription period on the product price setting screen.
- Added a new extension for String and `ProductFormDataModel` to format subscription period. The format follows the read-only version instead of Core for simplicity.
- Created a separate use case for the picker view for subscription period and interval.
- Updated the `TitleAndTextFieldTableViewCell` to accept custom input view.
- Updated the subscription period row to display the new picker view for inputting the subscription period and interval.
- Updated the completion handler for price setting to include subscription period and interval.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Enable the feature flag `subscriptionProducts` to be able to edit subscription products.
- Create a simple subscription product for your test site on wp-admin if you haven't already.
- Build the app, log in to your test store and navigate to the Products tab.
- Select a simple subscription product.
- Notice that the price row is displayed in place of the read-only subscription row. It should state the subscription period, like "$10 every 2 weeks".
- Tap the price row, notice a few changes:
  - A new row for subscription period is displayed below the Regular Price row. 
  - The Sale Price row is displayed in the second section with the new name Sale.
  - Tap on the Subscription Period row, notice that a picker is displayed for you to select the interval and period. 
  - Tap on Done on the toolbar above the picker, the new period should be displayed correctly.
  - Tap on Done on the price settings screen and tap Save on the product form, the new period should be updated correctly.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/4699750e-041b-4b60-90f3-9edea1f02a94" width=320 />



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Update SubscriptionPeriod to be decodable only

itsmeichigo committed:
Move sale price to the sale section

itsmeichigo committed:
Remove redundant error log when formatting sale price

itsmeichigo committed:
Show price row for subscription product if feature flag is enabled

itsmeichigo committed:
Add helper property to check for subscription product type

itsmeichigo committed:
Add new rows for subscription period on the price screen

itsmeichigo committed:
Add extension for subscription period description

itsmeichigo committed:
Unify subscription period info on the price screen

itsmeichigo committed:
Update regular price description for subscription product

itsmeichigo committed:
Update TitleAndTextFieldTableViewCell to support custom input view

itsmeichigo committed:
Update ProductPriceSettingsViewModel to update subscription period

itsmeichigo committed:
Update comment for localized string

itsmeichigo committed:
Make SubscriptionPeriod conform to CaseIterable

itsmeichigo committed:
Show picker for subscription period setting

itsmeichigo committed:
Fix failed tests

itsmeichigo committed:
Update subscription upon completion

itsmeichigo committed:
Fix linting errors

itsmeichigo committed:
Fix trailing space violation

itsmeichigo committed:
Update failed tests

itsmeichigo committed:
Update test for ProductVariationFormViewModel_UpdatesTests

itsmeichigo committed:
Make subscriptionPeriod and subscriptionPeriodInterval accessible from ProductPriceSettingsViewModel

itsmeichigo committed:
Preselect rows for subscription period picker

itsmeichigo committed:
Update ProductFormViewModel_ChangesTests

itsmeichigo committed:
Update ProductFormViewModel_UpdatesTests

itsmeichigo committed:
Update ProductVariationFormViewModel_ChangesTests

itsmeichigo committed:
Update ProductVariationFormViewModel_ObservablesTests and ProductFormViewModel_ObservablesTests

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift:
```diff
@@ -44,10 +44,15 @@ extension Product {
-        value = value
-            .replacingOccurrences(of: unit, with: "")
-            .replacingOccurrences(of: regexThousandSeparators, with: "$1", options: .regularExpression)
+        let value: String = {
+            guard let salePrice, salePrice.isNotEmpty else {
+                return ""
+            }
+            return (currencyFormatter.formatAmount(salePrice, with: unit) ?? "")
+                .replacingOccurrences(of: unit, with: "")
+                .replacingOccurrences(of: regexThousandSeparators, with: "$1", options: .regularExpression)
+        }()
```
Added this fix to avoid logging error when attempting to format an empty sale price.

itsmeichigo committed:
Handle completion instead of sending closure to avoid capturing self

itsmeichigo committed:
Add a separate use case for product subscription period

itsmeichigo committed:
Migrate to subscription period usecase

itsmeichigo committed:
Remove redundant methods

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewController.swift:
```diff
@@ -147,9 +164,10 @@ extension ProductPriceSettingsViewController {
-            onCompletion: { [weak self] (regularPrice, salePrice, dateOnSaleStart, dateOnSaleEnd, taxStatus, taxClass, hasUnsavedChanges) in
-                self?.onCompletion(regularPrice, salePrice, dateOnSaleStart, dateOnSaleEnd, taxStatus, taxClass, hasUnsavedChanges)
-            }, onError: { [weak self] error in
+            onCompletion: { [weak self] in
+                self?.onCompletion($0, $1, $2, $3, $4, $5, $6, $7, $8)
+            },
```
The shorthand is to avoid the verbosity of the parameters.

itsmeichigo committed:
Reuse logic for subscription period formatting

itsmeichigo committed:
Remove redundant import

itsmeichigo committed:
Dismiss input view on dragging the price form

itsmeichigo committed:
Update row title to Billing interval

itsmeichigo committed:
Fix failed tests

itsmeichigo committed:
Fix line length violations

itsmeichigo committed:
Remove redundant trailing spaces

itsmeichigo committed:
Update check for displaying subscription rows

itsmeichigo committed:
Fix failed unit test

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewController.swift:
```diff
@@ -45,6 +47,21 @@ final class ProductPriceSettingsViewController: UIViewController {
 
+    private lazy var subscriptionPeriodToolbar: UIToolbar = {
+        // Setting explicit frame size to avoid constraint conflicts.
+        let toolBar = UIToolbar(frame: .init(origin: .zero,
```
<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/5bb85bcd-7a77-48d8-bdb6-1781176ca5a7" width="320"/>

We might want to update the `Done` button color of the toolbar to match the top right Done button.

itsmeichigo replied:
Good catch! Updated in f1c2e44a28ef43f2d7253f99ae8c1bb73c2f4b1b.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewController.swift:
```diff
@@ -103,6 +120,7 @@ private extension ProductPriceSettingsViewController {
     func configureTableView() {
         tableView.rowHeight = UITableView.automaticDimension
         tableView.backgroundColor = .listBackground
+        tableView.keyboardDismissMode = .onDragWithAccessory
```
TIL 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewController.swift:
```diff
@@ -508,4 +558,20 @@ extension ProductPriceSettingsViewController {
+    enum Localization {
+        static let billingInterval = NSLocalizedString(
+            "productPriceSettingsViewController.billingIntervalRowTitle",
+            value: "Billing interval",
```
<img width="317" alt="Screenshot 2023-11-14 at 1 57 00 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/8117430f-7f51-438b-870a-2cf2212476a7">

FYI, The text from the PR's screen recording reads as "Subscription period". I hope we decided to stick to "Billing interval".

itsmeichigo replied:
My bad, I forgot to update the screencast in the PR description after updating the code. The description should be up-to-date now - thanks!

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewModel.swift:
```diff
@@ -116,10 +129,14 @@ final class ProductPriceSettingsViewModel: ProductPriceSettingsViewModelOutput {
+        var priceRows: [Row] = [.price]
+        if product.subscription != nil {
+            priceRows.append(contentsOf: [.subscriptionPeriod])
+        }
```
💡  Consider rewriting this into the following to avoid making `priceRows` mutable. This will help clarify that `priceRows` is not changed again. 
```suggestion
        let priceRows: [Row] = {
            guard product.subscription != nil else {
                return [.price]
            }
            return [.price, .subscriptionPeriod]
        }()
```

itsmeichigo replied:
Good point - updated in cb66979b8d3265a0958beda081ad09e9f5f55a1c.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductPriceSettingsViewModel.swift:
```diff
@@ -116,10 +129,14 @@ final class ProductPriceSettingsViewModel: ProductPriceSettingsViewModelOutput {
 
         // Sales section
-        var saleScheduleRows: [Row] = [.scheduleSale]
+        var saleScheduleRows: [Row] = [.salePrice, .scheduleSale]
```
❓  Is our idea to show salePrice always? I think showing `$0` as Sale price could be confusing. 

Update: I checked `trunk` and noticed that we should "Sale price" as zero there as well. Please feel free to keep it. 
<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/dd6f9268-3864-4621-86c7-5e04f024c238" width="320"/>

itsmeichigo replied:
Yup, I kept the behavior and only moved the row to a different section. I'd prefer to keep the existing behavior for now.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductSubscriptionPeriodPickerUseCase.swift:
```diff
@@ -0,0 +1,119 @@
+            return "\(row + 1)"
+        case Component.period.rawValue:
+            if pickerView.selectedRow(inComponent: Component.interval.rawValue) == 0 {
+                return SubscriptionPeriod.allCases[row].descriptionSingular
```
💡 What do you think about using `safe:` to read the array in this line and other lines in this file? Do you think it is not necessary as we are using `SubscriptionPeriod.allCases.count` to specify the `numberOfRowsInComponent`?

itsmeichigo replied:
I think it's better not to silence the error if it ever happens - we'll be alert when the crash occurs. Using `safe` doesn't crash the app but makes it tricky to spot when something goes wrong.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/ProductSubscriptionPeriodPickerUseCase.swift:
```diff
@@ -0,0 +1,119 @@
+            // reloads the period component to display the updated titles based on the interval.
+            pickerView.reloadComponent(Component.period.rawValue)
+        }
+        updateSubscriptionPeriodDescription()
```
❓ Do we want to update the values as soon as a selection is made in the picker view? Or do we want to wait for the user to tap `Done`? 

I noticed that we save the selection after the picker view selection is made. However, the table view row is not updated until the picker is dismissed by dragging down the table view.


https://github.com/woocommerce/woocommerce-ios/assets/524475/29498325-1a36-4c7f-ab3b-3bbf59f1a1b4

Should we update the table view row when the picker is changed (without waiting for the picker view to be dismissed)?

itsmeichigo replied:
I considered this, but updating the table view cell requires reloading the table view, which will then dismiss the picker, which can be annoying if the user is not done selecting yet. Letting them decide when to dismiss the picker is a better UX IMO.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormActionsFactory.swift:
```diff
@@ -303,9 +303,10 @@ private extension ProductFormActionsFactory {
 
         let actions: [ProductFormEditAction?] = [
-            .subscription(actionable: true),
+            editableSubscription ? .priceSettings(editable: true, hideSeparator: false) : .subscription(actionable: true),
```
❓ Should we avoid specifying `editable:` parameter as `true` and instead pass the `editable` value of `ProductFormActionsFactory`? 

It looks like that `ProductFormActionsFactory.editable` will be `false` for `.readonly` formType. 
```suggestion
            editableSubscription ? .priceSettings(editable: editable, hideSeparator: false) : .subscription(actionable: true),
```

itsmeichigo replied:
Nice catch 👍 Fixed in 43ad1116bd06a7e46312dcd9b4ec790e29dbb35f.

selanthiraiyan reviewed (approved):
Nice work, Huong! This works as expected. Thanks for adding/updating unit tests. 

I left a few non-blocking questions/comments.

itsmeichigo committed:
Update done button style for the picker toolbar

itsmeichigo committed:
Make priceRows immutable

itsmeichigo committed:
Pass editable to priceSettings action for subscription products

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews! I've addressed some of the issues and responded to your comments above. I'll turn on auto-merge now - but please feel free to follow up with any of the discussions above, I'll address any other issues in a different PR if necessary.
